### PR TITLE
Add RunPlugin support for python

### PR DESCRIPTION
**Pull request by Frassle**
# Description

Fills in RunPlugin for the python language host.

I've updated the various test component providers written in Python to use the `PulumiPlugin.yaml` and `__main__.py` instead of the shim .sh/.cmd files.

That is now two languages that have had to be disabled in `testComponentProviderSchema` because that doesn't use full plugin system for checking provider schemas. I think we should merge this as is, but I'll ensure that's fixed before we do NodeJS.

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: sdk/python/cmd/pulumi-language-python/main.go [hunk 1]
```diff
@@ -960,5 +959,62 @@ func (host *pythonLanguageHost) GetProgramDependencies(
...
+
+	cmd.Dir = req.Pwd
+	cmd.Env = req.Env
+	cmd.Stdout, cmd.Stderr = stdout, stderr
+
+	if virtualenv != "" {
+		cmd.Env = python.ActivateVirtualEnv(cmd.Env, virtualenv)
+	}
+	if err = cmd.Run(); err != nil {
+		if exiterr, ok := err.(*exec.ExitError); ok {
```
Comment (abhinav) [inline sdk/python/cmd/pulumi-language-python/main.go:998]: This should probably be an errors.As.
The cast will probably continue to work for the foreseeable future, but it's not required to.

```suggestion
		if exiterr := &exec.ExitError{}; errors.As(err, &exiterr) {
```

File: sdk/python/cmd/pulumi-language-python/main.go [hunk 2]
```diff
@@ -960,5 +959,62 @@ func (host *pythonLanguageHost) GetProgramDependencies(
...
+			if status, ok := exiterr.Sys().(syscall.WaitStatus); ok {
+				err = server.Send(&pulumirpc.RunPluginResponse{
+					Output: &pulumirpc.RunPluginResponse_Exitcode{Exitcode: int32(status.ExitStatus())},
+				})
+			} else {
+				err = fmt.Errorf("program exited unexpectedly: %w", exiterr)
+			}
+		} else {
+			return fmt.Errorf("problem executing plugin program (could not run language executor): %w", err)
+		}
```
Comment (abhinav) [inline sdk/python/cmd/pulumi-language-python/main.go:1008]: optional:
We're returning the error we're building above in all cases.
We can drop the else clauses if we return as we build these errors.

```suggestion
				return server.Send(&pulumirpc.RunPluginResponse{
					Output: &pulumirpc.RunPluginResponse_Exitcode{Exitcode: int32(status.ExitStatus())},
				})
			}
			return fmt.Errorf("program exited unexpectedly: %w", exiterr)
		}
		return fmt.Errorf("problem executing plugin program (could not run language executor): %w", err)
```

File: sdk/python/cmd/pulumi-language-python/main.go [hunk 3]
```diff
@@ -960,5 +959,62 @@ func (host *pythonLanguageHost) GetProgramDependencies(
...
+	var err error
+	var virtualenv string
+	if host.virtualenv != "" {
+		virtualenv = host.virtualenvPath
+		if !python.IsVirtualEnv(virtualenv) {
+			return python.NewVirtualEnvError(host.virtualenv, virtualenv)
+		}
+		cmd = python.VirtualEnvCommand(virtualenv, "python", args...)
+	} else {
+		cmd, err = python.Command(server.Context(), args...)
```
Comment (abhinav) [inline sdk/python/cmd/pulumi-language-python/main.go:977]: optional: err is needed only for the else block. Code down below gets its own err.
Can move the declaration closer to where it's used, reducing its scope.
```suggestion
	var virtualenv string
	if host.virtualenv != "" {
		virtualenv = host.virtualenvPath
		if !python.IsVirtualEnv(virtualenv) {
			return python.NewVirtualEnvError(host.virtualenv, virtualenv)
		}
		cmd = python.VirtualEnvCommand(virtualenv, "python", args...)
	} else {
		var err error
		cmd, err = python.Command(server.Context(), args...)
```

That makes the reading flow of this code block nicer.
Like, `var X, Y; if { ... } else { ... }` suggests that X and Y are outputs of that "paragraph" of code.
In this case, only cmd and virtualenv are outputs of this block.

File: sdk/python/cmd/pulumi-language-python/main.go [hunk 4]
```diff
@@ -960,5 +959,62 @@ func (host *pythonLanguageHost) GetProgramDependencies(
...
+			}
+		} else {
+			return fmt.Errorf("problem executing plugin program (could not run language executor): %w", err)
+		}
+	}
+
+	if err != nil {
+		return err
+	}
+
```
Comment (abhinav) [inline sdk/python/cmd/pulumi-language-python/main.go:1014]: Can delete this if we return above as needed.

```suggestion
```

**Review by abhinav — APPROVED** [review]

File: tests/integration/integration_python_test.go [hunk 5]
```diff
@@ -963,6 +963,11 @@ func TestPythonTranslation(t *testing.T) {
 }
 
 func TestComponentProviderSchemaPython(t *testing.T) {
+	// We no longer have shim files so there's no native binary for the testComponentProviderSchema to just
+	// exec. It _ought_ to be rewritten to use the plugin host framework so that it starts the component up
+	// the same as all the other tests are doing (via shimless).
+	t.Skip("testComponentProviderSchema needs to be updated to use a plugin host to deal with non-native-binary providers")
```
[the remark this answers is not in the document]
Comment (Frassle) [reply tests/integration/integration_python_test.go:970]: Created and linked

**Comment by Frassle** [comment]
bors merge

**Pull request merged**